tests/gcp/test-utils_gcp.R

```diff
@@ -0,0 +1,194 @@
+  projectId <- "gcp-setup-demo"
+
+  googleCloudStorageR::gcs_create_bucket(bucket, projectId = projectId)
+  on.exit(gcp_gcs_delete_bucket(bucket))
+  expect_false(gcp_gcs_exists(key = "x", bucket = bucket))
+  tmp <- tempfile()
+  writeLines("x", tmp)
+  googleCloudStorageR::gcs_upload(tmp, bucket = bucket, name = "x")
+  expect_true(gcp_gcs_exists(key = "x", bucket = bucket))
+
+  head <- gcp_gcs_head(key = "x", bucket = bucket)
+  expect_true(inherits(head, "gcs_objectmeta"))
+  expect_true(is.character(head$etag))
+  expect_true(nzchar(head$etag))
+})
+
+tar_test("gcp_gcs_download()", {
+  skip_if_no_gcp()
+  auth_gcp()
+  bucket <- random_bucket_name()
+  # needs to be a GCP project the tester auth has access to
+  projectId <- "gcp-setup-demo"
+
+  googleCloudStorageR::gcs_create_bucket(bucket, projectId = projectId)
+  on.exit(gcp_gcs_delete_bucket(bucket))
+  expect_false(gcp_gcs_exists(key = "x", bucket = bucket))
+  tmp <- tempfile()
+  writeLines("x", tmp)
+  googleCloudStorageR::gcs_upload(tmp, bucket = bucket, name = "x")
+  tmp2 <- tempfile()
+  expect_false(file.exists(tmp2))
+  gcp_gcs_download(file = tmp2, key = "x", bucket = bucket)
+  expect_equal(readLines(tmp2), "x")
+})
+
+tar_test("gcp_gcs_upload() without headers", {
+  skip_if_no_gcp()
+  auth_gcp()
+  bucket <- random_bucket_name()
+  # needs to be a GCP project the tester auth has access to
+  projectId <- "gcp-setup-demo"
+
+  googleCloudStorageR::gcs_create_bucket(bucket, projectId = projectId)
+  on.exit(gcp_gcs_delete_bucket(bucket))
+
+  expect_false(gcp_gcs_exists(key = "x", bucket = bucket))
+  tmp <- tempfile()
+  writeLines("x", tmp)
+  gcp_gcs_upload(
+    file = tmp,
+    key = "x",
+    bucket = bucket
+  )
+  expect_true(gcp_gcs_exists(key = "x", bucket = bucket))
+})
+
+tar_test("gcp_gcs_upload() and download with metadata", {
+  skip_if_no_gcp()
+  auth_gcp()
+  bucket <- random_bucket_name()
+  # needs to be a GCP project the tester auth has access to
+  projectId <- "gcp-setup-demo"
+
+  googleCloudStorageR::gcs_create_bucket(bucket, projectId = projectId)
+  on.exit(gcp_gcs_delete_bucket(bucket))
+
+  expect_false(gcp_gcs_exists(key = "x", bucket = bucket))
+  tmp <- tempfile()
+  writeLines("x", tmp)
+  gcp_gcs_upload(
+    file = tmp,
+    key = "x",
+    bucket = bucket,
+    metadata = list("custom" = "custom_metadata")
+  )
+  expect_true(gcp_gcs_exists(key = "x", bucket = bucket))
+  head <- gcp_gcs_head(key = "x", bucket = bucket)
+  expect_equal(head$metadata$custom, "custom_metadata")
+  tmp2 <- tempfile()
+  expect_false(file.exists(tmp2))
+  gcp_gcs_download(file = tmp2, key = "x", bucket = bucket)
+  expect_equal(readLines(tmp2), "x")
+})
+
+tar_test("gcp_gcs upload twice, get the correct version", {
+  skip_if_no_gcp()
+  auth_gcp()
+  bucket <- random_bucket_name()
+  # needs to be a GCP project the tester auth has access to
+  projectId <- "gcp-setup-demo"
+  googleCloudStorageR::gcs_create_bucket(bucket,
```

Comment:
Another place where it would be nice to give each argument its own indented line.